lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
I'm wary of putting an ``until`` here, as it's not guaranteed that you'll eventually read the reply with a ``response_to`` matching the last message's ``request_id``. What I would do instead is pass the expected ``request_id`` to #read like this:

```ruby
messages.last.replyable? ? read(messages.last.request_id) : nil
```

then I'd change the method signature of Connectable#read to this

```ruby
      def read(request_id = nil)
        ensure_connected do |socket|
          Protocol::Reply.deserialize(socket, max_message_size, request_id)
        end
      end
```

and then the method signature of Message.deserialize to this:

```ruby
def self.deserialize(io, max_message_size = MAX_MESSAGE_SIZE, expected_request_id = nil)
```

and raise an Exception in the same way we're raising one for the length:

```ruby
if expected_request_id && expected_request_id != response_to
  raise Error::UnexpectedResponse.new(expected_request_id)
end
```

What do you think? We'll address the scenario in which a socket has a killed-thread's unread data by this ticket:
https://jira.mongodb.org/browse/RUBY-1118 so what the until loop accomplishes wouldn't be necessary.